cluster-autoscaler/simulator/basic_cluster_snapshot.go

```diff
@@ -0,0 +1,263 @@
+}
+
+func (data *internalBasicSnapshotDataNodeLister) HavePodsWithAffinityList() ([]*schedulernodeinfo.NodeInfo, error) {
+	havePodsWithAffinityList := make([]*schedulernodeinfo.NodeInfo, 0, len(data.nodeInfoMap))
+	for _, v := range data.nodeInfoMap {
+		if len(v.PodsWithAffinity()) > 0 {
+			havePodsWithAffinityList = append(havePodsWithAffinityList, v)
+		}
+	}
+	return havePodsWithAffinityList, nil
+}
+
+func (data *internalBasicSnapshotDataNodeLister) Get(nodeName string) (*schedulernodeinfo.NodeInfo, error) {
+	if v, ok := data.nodeInfoMap[nodeName]; ok {
+		return v, nil
+	}
+	return nil, fmt.Errorf("node %s not in snapshot", nodeName)
+}
+
+func (data *internalBasicSnapshotDataPodLister) List(selector labels.Selector) ([]*apiv1.Pod, error) {
+	alwaysTrue := func(p *apiv1.Pod) bool { return true }
+	return data.FilteredList(alwaysTrue, selector)
+}
+
+func (data *internalBasicSnapshotDataPodLister) FilteredList(podFilter schedulerlisters.PodFilter, selector labels.Selector) ([]*apiv1.Pod, error) {
+	pods := make([]*apiv1.Pod, 0)
+	for _, n := range data.nodeInfoMap {
+		for _, pod := range n.Pods() {
+			if podFilter(pod) && selector.Matches(labels.Set(pod.Labels)) {
+				pods = append(pods, pod)
+			}
+		}
+	}
+	return pods, nil
+}
+
+func (data *internalBasicSnapshotData) Pods() schedulerlisters.PodLister {
+	return (*internalBasicSnapshotDataPodLister)(data)
+}
+
+func (data *internalBasicSnapshotData) NodeInfos() schedulerlisters.NodeInfoLister {
+	return (*internalBasicSnapshotDataNodeLister)(data)
+}
+
+// NewEmptySnapshot initializes a Snapshot struct and returns it.
+func newInternalBasicSnapshotData() *internalBasicSnapshotData {
+	return &internalBasicSnapshotData{
+		nodeInfoMap: make(map[string]*schedulernodeinfo.NodeInfo),
+	}
+}
+
+func (data *internalBasicSnapshotData) clone() *internalBasicSnapshotData {
+	clonedNodeInforMap := make(map[string]*schedulernodeinfo.NodeInfo)
+	for k, v := range data.nodeInfoMap {
+		clonedNodeInforMap[k] = v.Clone()
+	}
+	return &internalBasicSnapshotData{
+		nodeInfoMap: clonedNodeInforMap,
+	}
+}
+
+func (data *internalBasicSnapshotData) addNode(node *apiv1.Node) error {
+	if _, found := data.nodeInfoMap[node.Name]; found {
+		return fmt.Errorf("node %s already in snapshot", node.Name)
+	}
+	nodeInfo := schedulernodeinfo.NewNodeInfo()
+	err := nodeInfo.SetNode(node)
+	if err != nil {
+		return fmt.Errorf("cannot set node in NodeInfo; %v", err)
+	}
+	data.nodeInfoMap[node.Name] = nodeInfo
+	return nil
+}
+
+func (data *internalBasicSnapshotData) removeNode(nodeName string) error {
+	if _, found := data.nodeInfoMap[nodeName]; !found {
+		return fmt.Errorf("node %s not in snapshot", nodeName)
+	}
+	delete(data.nodeInfoMap, nodeName)
+	return nil
+}
+
+func (data *internalBasicSnapshotData) addPod(pod *apiv1.Pod, nodeName string) error {
+	if _, found := data.nodeInfoMap[nodeName]; !found {
+		return fmt.Errorf("node %s not in snapshot", nodeName)
+	}
+	data.nodeInfoMap[nodeName].AddPod(pod)
```

Comment:
Would it make sense to set pod.Spec.NodeName here if it's not already set?

Reply:
I would rather not mutate the `Pod` objects which are given to us - unless we make clones of those.
And cloning those can be somewhat costly.

If we clone those we should ensure that NodeName in pod matches the node it is scheduled to. Otherwise I do not believe so. 
Especially logic like set-if-not set would be cumbersome, and lead to unexpected state.

Reply:
That's a fair point. If scheduler is not validating that internally, it's probably an overkill to deal with that.